internal/engine/ingester/deps/deps.go

```diff
@@ -0,0 +1,213 @@
+			return nil, fmt.Errorf("%w: %s: branch %s", engerrors.ErrEvaluationFailed,
+				provifv1.ErrProviderGitBranchNotFound, branch)
+		} else if errors.Is(err, provifv1.ErrRepositoryEmpty) {
+			return nil, fmt.Errorf("%w: %s", engerrors.ErrEvaluationSkipped, provifv1.ErrRepositoryEmpty)
+		}
+		return nil, err
+	}
+
+	wt, err := r.Worktree()
+	if err != nil {
+		return nil, fmt.Errorf("could not get worktree: %w", err)
+	}
+
+	deps, err := scanFs(ctx, wt.Filesystem)
+	if err != nil {
+		return nil, fmt.Errorf("could not scan filesystem: %w", err)
+	}
+
+	head, err := r.Head()
+	if err != nil {
+		return nil, fmt.Errorf("could not get head: %w", err)
+	}
+
+	hsh := head.Hash()
+
+	chkpoint := checkpoints.NewCheckpointV1Now().
+		WithBranch(branch).
+		WithCommitHash(hsh.String())
+
+	return &interfaces.Result{
+		Object:     deps,
+		Checkpoint: chkpoint,
+	}, nil
+}
+
+func (gi *Deps) getBranch(repo *pb.Repository, branch string) string {
+	// If the user has specified a branch, use that
+	if branch != "" {
+		return branch
+	}
+
+	// If the branch is provided in the rule-type
+	// configuration, use that
+	if gi.cfg.GetRepo().Branch != "" {
+		return gi.cfg.GetRepo().Branch
+	}
+	if repo.GetDefaultBranch() != "" {
+		return repo.GetDefaultBranch()
+	}
+
+	// If the branch is not provided in the rule-type
+	// configuration, use the default branch
+	return defaultBranch
+}
+
+func scanFs(ctx context.Context, memFS billy.Filesystem) (*sbom.NodeList, error) {
+	// have to down-cast here, because scalibr needs multiple io/fs types
+	wrapped, ok := iofs.New(memFS).(scalibr_fs.FS)
+	if !ok {
+		return nil, fmt.Errorf("error converting filesystem to ReadDirFS")
+	}
+
+	desiredCaps := scalibr_plugin.Capabilities{
+		OS:            scalibr_plugin.OSLinux,
+		Network:       true,
+		DirectFS:      false,
+		RunningSystem: false,
+	}
+
+	scalibrFs := scalibr_fs.ScanRoot{FS: wrapped}
+	scanConfig := scalibr.ScanConfig{
+		ScanRoots: []*scalibr_fs.ScanRoot{&scalibrFs},
+		// All includes Ruby, Dotnet which we're not ready to test yet, so use the more limited Default set.
+		FilesystemExtractors: list.FilterByCapabilities(list.Default, &desiredCaps),
+		Capabilities:         &desiredCaps,
+	}
+
+	scanner := scalibr.New()
+	scanResults := scanner.Scan(ctx, &scanConfig)
+
+	if scanResults == nil || scanResults.Status == nil {
+		return nil, fmt.Errorf("error scanning files: no results")
+	}
+	if scanResults.Status.Status != scalibr_plugin.ScanStatusSucceeded {
+		return nil, fmt.Errorf("error scanning files: %s", scanResults.Status)
+	}
+
+	res := sbom.NewNodeList()
+	for _, inv := range scanResults.Inventories {
+		node := &sbom.Node{
+			Type:    sbom.Node_PACKAGE,
+			Id:      uuid.New().String(),
+			Name:    inv.Name,
+			Version: inv.Version,
+			Identifiers: map[int32]string{
+				int32(sbom.SoftwareIdentifierType_PURL): inv.Extractor.ToPURL(inv).String(),
+				// TODO: scalibr returns a _list_ of CPEs, but protobom will store one.
+				// use the first?
```

Comment:
Ideally the most specific one. Now that package URL has official support for ranges, the one-purl-per-package rule is broken for purls. This means that perhaps we should create a new identifier list to capture more than one.

Reply:
I'm slightly confused about what you're suggesting here.  I can see why PURLs supporting ranges would be useful _in general_, but I'm not sure if you're suggesting:

1. That upstream `protobom` needs a change in the `identifiers` type.
2. That Scalibr needs a change in the `ToPURL` function.
3. That this code should be parsing the purl / Inventory and creating multiple `Node` objects when the Inventory contains certain types of data.

Looking at the scalibr code, it looks like CPEs are only reported from the SPDX and CDX extractors, so multiple (or > 0) CPEs seems moot for our current use cases.  :puzzled:

Reply:
I meant number 1, but there are use cases when you may also want to return purls with ranges, more than one purl, and/or more than one CPE. It all depends on what you are trying to match.